web/models/setting/system_variable.ex

```diff
@@ -0,0 +1,96 @@
+defmodule CoursePlanner.SystemVariable do
+  @moduledoc """
+  This module holds the model for one system variable
+  """
+  use CoursePlanner.Web, :model
+
+  alias Ecto.Changeset
+
+  schema "system_variables" do
+    field :key,      :string
+    field :value,    :string
+    field :type,     :string
+    field :visible,  :boolean
+    field :editable, :boolean
+
+    timestamps()
+  end
+
+  @doc """
+  Builds a changeset based on the `struct` and `params`.
+  """
+  def changeset(struct, params \\ %{}) do
+    target_params =
+      [
+       :key,
+       :value,
+       :type,
+       :visible,
+       :editable
+      ]
+
+    struct
+    |> cast(params, target_params)
+    |> validate_required(target_params)
+    |> validate_value_type()
+  end
+
+  def changeset(struct, params, :update) do
+    struct
+    |> changeset(params)
+    |> validate_editable()
+  end
+
+  defp validate_editable(%{valid?: true} = changeset) do
+    editable = changeset |> Changeset.get_field(:editable)
+
+    if editable do
+      changeset
+    else
+      add_error(changeset, :value, "operation is forbidden")
+    end
+  end
+  defp validate_editable(changeset), do: changeset
+
+  defp validate_value_type(%{valid?: true} = changeset) do
+    value = changeset |> Changeset.get_field(:value)
+    type  = changeset |> Changeset.get_field(:type)
+
+    validation_result =
+      case type do
+        "string"  -> {:ok, value}
+        "integer" -> parse_integer(value)
+        "boolean" -> parse_boolean(value)
+        _         -> {:error, "unknown type"}
+      end
+
+    case validation_result do
+      {:ok, _} -> changeset
+      {:error, message} -> add_error(changeset, :value, message)
+    end
+  end
+  defp validate_value_type(changeset), do: changeset
+
+  def parse_integer(value) do
+    case Integer.parse(value) do
+      {parsed_value, ""} -> validate_number_rage_inclusive(parsed_value, 0, 1_000_000)
```

Comment:
But "validate_number_rage_inclusive" adds the error correctly itself. converting to your solution then I have to include an add_error instead of validate_number_rage_inclusive which to me both are the same